### PR TITLE
Issue 41617: Chromatograms are not displayed for precursors that do not have any quantitative transitions

### PR DESCRIPTION
**Pull request by vagisha**
#### Rationale
If a precursor does not have any quantitative transitions the chromatogram chart does not display any peaks.

#### Changes
- Render non-quantitative precursor and transition peaks as dashed lines
- Use the start and end times for a precursor's transition peaks if the minStartTime and minEndTime on the precursor peak are not set

### CODE REVIEW

File: src/org/labkey/targetedms/chart/ChromatogramDataset.java [hunk 1]
```diff
@@ -1155,9 +1206,8 @@ public int compare(TransChromInfoPlusTransition t1, TransChromInfoPlusTransition
         }
     }
 
-    static class MoleculeTransChromInfoPlusTransition
+    static class MoleculeTransChromInfoPlusTransition extends TransitionChromInfoPlusGeneralTransition<MoleculeTransition>
     {
-        private TransitionChromInfo _transChromInfo;
         private MoleculeTransition _transition;
```
Comment (labkey-jeckels) [inline src/org/labkey/targetedms/chart/ChromatogramDataset.java]: Could push this member variable up to the abstract superclass and make getTransition() non-abstract there.

File: src/org/labkey/targetedms/query/PrecursorManager.java [hunk 2]
```diff
@@ -809,6 +821,44 @@ public static boolean hasOptimizationPeaks(long peptideId)
...
+        if(pci == null)
+        {
+            return null;
+        }
+        if(pci.getMaxEndTime() != null)
+        {
+            return pci.getMaxEndTime();
+        }
+
+        // Issue 41617: Chtomatograms are not displayed for precursors that do not have any quantitative transitions
```
Comment (labkey-jeckels) [inline src/org/labkey/targetedms/query/PrecursorManager.java]:
```suggestion
        // Issue 41617: Chromatograms are not displayed for precursors that do not have any quantitative transitions
```

File: src/org/labkey/targetedms/query/PrecursorManager.java [hunk 3]
```diff
@@ -809,6 +821,44 @@ public static boolean hasOptimizationPeaks(long peptideId)
...
+        if(pci == null)
+        {
+            return null;
+        }
+        if(pci.getMinStartTime() != null)
+        {
+            return pci.getMinStartTime();
+        }
+
+        // Issue 41617: Chtomatograms are not displayed for precursors that do not have any quantitative transitions
```
Comment (labkey-jeckels) [inline src/org/labkey/targetedms/query/PrecursorManager.java]:
```suggestion
        // Issue 41617: Chromatograms are not displayed for precursors that do not have any quantitative transitions
```

**Review by labkey-jeckels — APPROVED** [review]
A couple of small items to consider, but up to you.

A bigger change would be to combine the queries that get the min and max values for the retention times. It looks like we always want both of them, and it would be faster to do a single query. But I don't think this will be a significant issue in practice because we're not likely to need the min/max for a lot of different transitions or precursors, correct?

**Comment by vagisha** [comment]
Queries to get the min and max retention times from the transitionchrominfo table have been combined into a single query.

**Review by labkey-jeckels — APPROVED** [review]
Changes look good!

**Pull request merged**